Redact IDs from urls for GA

Rosa-Fox opened the pull request:
What
----

This will redact any value in the query string (`window.location.search`) parameters, but keep the key present - this ensures that Google Analytics knows about the keys, but not the values of those keys. This will prevent any personally identifiable information (PII) from leaking into GA.

For example:

`https://example.com/?really_important_thing=top_secret` 

will be reported as 

`https://example.com/?really_important_thing=<REALLY_IMPORTANT_THING>`

We remove the `?` from `window.location.search`, and then `split()` that by the ampersands. We then split the `key=value` into a key and value. 

The query string is then put together without the value - `key=<KEY>` - and the whole URL is constructed again to be returned as a string for reporting into Google Analytics.

Jasmine tests have been added as part of this pull request - `bundle exec rake jasmine:ci` to run them, or `bundle exec rake` to run all tests.

## Why

Because PII in GA is bad and this will help prevent any PII from being reported.

How to review
-------------

- The Google doc linked to the Trello card may also be useful for giving some more context
- Check the Jasmine tests cover all bases

Links
-----

[Trello](https://trello.com/c/J5W2oQmv/569-dont-send-google-analytics-user-transaction-id-query-parameters-for-vulnerable-people-form)

Rosa-Fox committed:
Redact IDs from urls for GA

- Technically this isn't personally identifiable information (PII) on its own, but it could be along with the backend data.
- Jasmine tests will be added in a future PR.

leenagupte commented on app/assets/javascripts/analytics.js.erb:
```diff
@@ -1,7 +1,21 @@
+  url = url.replace('true', 'contact_gp')
+  url = url.replace('false', 'contact_gp')
+
+  if(url.includes('reference_number') || 'contact_gp') {
+    return url
+  } else {
+    return location.protocol + '//' + location.host + location.pathname
+  }
```
I don't think I understand why this `if/else` is needed. What happens if the function always returns `url`?

Rosa-Fox replied:
This was a result of the conversation [here](https://github.com/alphagov/govuk-coronavirus-business-volunteer-form/pull/361). Originally I just returned `url` but using `if/else` was cautionary incase other params are added and we miss changing this code. Have realised its still a bit flawed in that if only one of `reference_number` or `contact_gp` are matched then the other is not redacted, will change to an `&&`...

leenagupte replied:
I didn't realise that we're trying to avoid sending other params that could be in the query-string to GA too.  Do you think it's possible to loop through all/any of the query params and replace them all with "XXXXXXX"? It means that if `contact_gp` for example is ever renamed, we don't need to worry about having to update it here too.

leenagupte replied:
I've fallen down a rabbit hole. Apparently there's a `URLSearchParams` class that gets query params and interrogates query params: https://flaviocopes.com/urlsearchparams/

Rosa-Fox replied:
nice! Thanks, I will try that :)

Rosa-Fox replied:
@leenagupte thanks for suggested, is a much stronger approach. I have amended the commit.

leenagupte commented on app/assets/javascripts/analytics.js.erb:
```diff
@@ -2,6 +2,14 @@
+  for (const [key, value] of params.entries()) {
+    params.set(key, "XXXX")
+  }
+  return location.protocol + '//' + location.host + location.pathname + params
```
```suggestion
  return location.protocol + '//' + location.host + location.pathname + '?' + params
```

Without the '?' the url returned looks like: 
```
"http://localhost:5000/confirmationcontact_gp=XXXX&reference_number=XXXX"
```

I'm not sure if it's ok to add a trailing `?` to urls if there aren't any querystring params. It might be worth returning `location.protocol + '//' + location.host + location.pathname` if params is `null` before doing the loop?

Rosa-Fox replied:
ahh! have amended :) @leenagupte

leenagupte reviewed (commented):
This is looking good. I'm excited that `URLSearchParams` works!

I've spotted one small thing in the url generated, otherwise it looks good.

I think you definitely should get someone how knows Javascript to look this over as well though.

leenagupte reviewed (dismissed):
💯 👍 🎉

alex-ju commented on app/assets/javascripts/analytics.js.erb:
```diff
@@ -2,6 +2,25 @@
 new GOVUK.Modules.CookieBanner().start($(element))
 
+var redactPII = function() {
+  const params = new URLSearchParams(window.location.search)
```
`URLSearchParams` is not supported in Internet Explorer (11 and below) so we can't rely on it, unfortunately.

alex-ju commented on app/assets/javascripts/analytics.js.erb:
```diff
@@ -2,6 +2,25 @@
 
+var redactPII = function() {
+  const params = new URLSearchParams(window.location.search)
+  for (const [key, value] of params.entries()) {
```
We don't use `const` and `let`, only `var` as we don't transpile from ES6 and the new declarations are not supported in Internet Explorer (in v11 not supported in for-in and for-of loops and below v11 there is no support whatsoever).

alex-ju reviewed (commented):
Looks good, but unfortunately this approach doesn't cover all the supported browsers so if we must do it, it will have to be regex. I would be cautious about adding complexity to our scripts (that are served to every visitor) and given this isn't technically PII I would suggest we exclude them on the GA-side using 'Exclude Query Parameters'.

injms committed:
Add Jasmine and configure it to test JavaScripts

Copy of work done in alphagov/govuk-coronavirus-business-volunteer-form#389

Co-authored-by: Graham Lewis <graham.lewis@digital.cabinet-office.gov.uk>

injms committed:
Add stripPII function

Add function to redact the `reference_number` and `contact_gp` parameters in
the URL query string so that it can be used with Google Analytics cross domain
tracking.

This searches the query string (aka `window.location.search`) for the keys
`reference_number` and `contact_gp`, and redacts their values. This ensures
that no other information gets lost from the query string.

Copy of work done in alphagov/govuk-coronavirus-business-volunteer-form#389

Co-authored-by: Graham Lewis <graham.lewis@digital.cabinet-office.gov.uk>

injms committed:
Add stripPII function to analytics

 - stripPII function redacts *all* query string parameters
 - Adds the stripPII function to the application.js file
 - Sets up stripPII function on the `window.GOVUK` object
 - Remove the previous stripPII function and add the newer stripPII function
   to the analytics script.

Based on work done in alphagov/govuk-coronavirus-business-volunteer-form#389

Co-authored-by: Graham Lewis <graham.lewis@digital.cabinet-office.gov.uk>

Rosa-Fox commented:
This looks great, thanks so much @injms! Thanks for taking me through the code, I have tested the updated changes in the browser and run locally... but I can't approve as I originally opened them PR.

injms commented:
> This looks great, thanks so much @injms! Thanks for taking me through the code, I have tested the updated changes in the browser and run locally... but I can't approve as I originally opened them PR.

Ah, okay - makes sense. @alex-ju, could you review this again please? 🙏

alex-ju reviewed (approved):
Re-checked. It's aligned with https://github.com/alphagov/govuk-coronavirus-business-volunteer-form/pull/389, happy to get it in 👏
